File: echo-guardian-production-foundation-v1.0/native/ios/EchoGuardianKit/Sources/EchoGuardianKit/Runtime/NativeCoreCycleBridge.swift
import Foundation

public struct NativeCoreCycleBridgeResult: Codable, Equatable {
    public let schemaVersion: String
    public let observationJson: String
    public let rawSensorRetained: Bool
    public let privateSpace: Bool
    public let readyForCore: Bool
    public let plainLanguageSummary: String
}

public enum NativeCoreCycleBridgeError: Error {
    case privateSpaceBlocked
    case rawLikeMetadataBlocked
    case encodingFailed
}

public final class NativeCoreCycleBridge {
    public init() {}

    public func bridge(observation: MinimizedSensorObservation) throws -> NativeCoreCycleBridgeResult {
        if observation.privateSpace || observation.spaceId == "bathroom" || observation.spaceId == "bedroom" {
            throw NativeCoreCycleBridgeError.privateSpaceBlocked
        }
        if observation.rawSensorRetained {
            throw NativeCoreCycleBridgeError.rawLikeMetadataBlocked
        }
        let rawLike = ["raw", "raw_data", "raw_sample", "samples", "audio_buffer", "video_frame", "transcript", "waveform", "packet_capture"]
        if observation.derivedMetadata.keys.contains(where: { rawLike.contains($0.lowercased()) }) {
            throw NativeCoreCycleBridgeError.rawLikeMetadataBlocked
        }
        let data = try JSONEncoder().encode(observation)
        guard let json = String(data: data, encoding: .utf8) else { throw NativeCoreCycleBridgeError.encodingFailed }
        return NativeCoreCycleBridgeResult(
            schemaVersion: "0.6",
            observationJson: json,
            rawSensorRetained: false,
            privateSpace: false,
            readyForCore: true,
            plainLanguageSummary: "The native observation is minimized and ready for the Echo Guardian core loop."
        )
    }
}

File: echo-guardian-production-foundation-v1.0/native/android/echo-guardian-core/src/main/java/org/echoguardian/runtime/NativeCoreCycleBridge.kt
package org.echoguardian.runtime

import org.echoguardian.sensors.MinimizedSensorObservation

class NativeCoreCycleBridge {
    private val forbiddenKeys = setOf("raw", "raw_data", "raw_sample", "samples", "audio_buffer", "video_frame", "transcript", "waveform", "packet_capture")

    fun bridge(observation: MinimizedSensorObservation): Map<String, Any> {
        require(!observation.privateSpace && observation.spaceId != "bathroom" && observation.spaceId != "bedroom") {
            "Private-space monitoring is blocked by default in v0.6"
        }
        require(!observation.rawSensorRetained) { "Raw sensor retention is blocked by default in v0.6" }
        require(observation.derivedMetadata.keys.none { forbiddenKeys.contains(it.lowercase()) }) {
            "Raw-like metadata keys are blocked by default in v0.6"
        }
        return mapOf(
            "schema_version" to "0.6",
            "raw_sensor_retained" to false,
            "private_space" to false,
            "ready_for_core" to true,
            "plain_language_summary" to "The native observation is minimized and ready for the Echo Guardian core loop."
        )
    }
}
